Restore basic functionality to the bytecode debugger

damiendoligez opened the pull request:
~~This is work in progress.~~ I still need to figure out what must be done to handle backtraces in the presence of fibers.
[edit: now it's ready]

damiendoligez committed:
debugger: the main program is now code fragment number 3 instead of 0

damiendoligez committed:
debugger: fix locking issues with checkpoints and I/O operations

xavierleroy commented on runtime/debugger.c:
```diff
@@ -246,6 +246,16 @@ void caml_debugger_init(void)
     sock_addr.s_inet.sin_port = htons(atoi(port));
     sock_addr_len = sizeof(sock_addr.s_inet);
   }
+  /* We completely disable channel locking because:
+     1. The debugger is not compatible with threads, so we don't need locks.
+     2. This file doesn't use locks around channel operations, but
+        function [check_pending] in io.c assumes that the channel is locked
+        and tries to unlock it.
+  */
+  caml_channel_mutex_free = NULL;
+  caml_channel_mutex_lock = NULL;
+  caml_channel_mutex_unlock = NULL;
+  caml_channel_mutex_unlock_exn = NULL;
```
I agree this should not make any difference for a single-domain single-threaded program.  Yet, I'd prefer if you would just take the locks on `dbg_in` and `dbg_out` just after they are created, and never release these locks.

(Note to self: investigate why `caml_ml_close_channel` does not lock the channel .)

damiendoligez replied:
Ah yes, that's a much better solution, thanks.

OlivierNicole commented:
I was also looking at this, but you are more advanced. If I can help with reviews or testing, let me know.

damiendoligez committed:
debugger: keep debugger IO channels locked at all times

xavierleroy commented on runtime/debugger.c:
```diff
@@ -138,10 +138,16 @@ static void open_connection(void)
+     private to this code and never used concurrently. */
+  Lock (dbg_in);
+  Lock (dbg_out);
+  /* We also need to remove them from the list of all open channels,
+     otherwise the flush-everything-at-exit code will try to relock them. */
```
I'm surprised.  `caml_open_descriptor_{in,out}` does not set the `CHANNEL_FLAG_MANAGED_BY_GC` flag, so `caml_ml_out_channels_list` should not include these channels in its result list.  What's going on?

damiendoligez replied:
I don't see any mention of `CHANNEL_FLAG_MANAGED_BY_GC` in `caml_{link,unlink}_channel`. The list of all open channels is used in `caml_ml_out_channels_list`, which is used to implement `Stdlib.flush_all`, which is itself used in `at_exit` to flush everything when the program exits.

damiendoligez replied:
Oh I missed that. It's a strange interaction between memory management and `flush_all`.

xavierleroy replied:
Well, there's the channels created from OCaml, which are subject to finalization and closed automatically when the program exits, and there's the channels created by the runtime system for its own uses (loading data from bytecode executables; communicating with the debugger), which are neither finalized nor auto-closed.  So it's not absurd to use a single flag `MANAGED_BY_GC` to govern both finalization and `flush_all`/`close_all`...

damiendoligez committed:
The debugger does not need to unlink the channels because they don't have the CHANNEL_FLAG_MANAGED_BY_GC flag.

damiendoligez committed:
update dependencies

xavierleroy commented:
This looks rather good to me.  Why don't you remove the Draft status so that it can get reviewed properly?

xavierleroy commented:
Ping.  Why don't you remove the Draft status so that it can get reviewed properly?

damiendoligez committed:
debugger: update to handle stack backtrace in the presence of fibers

damiendoligez committed:
Fix embedded file name (and line number) for stdlib.ml.in to get error
messages right and install it so the debugger can step through it.

damiendoligez commented:
Stack backtrace is now working. The `next` instruction still behaves a little strange around fibers, but the debugger is already usable. I'll do another PR for `next` but I think it's less urgent.

xavierleroy committed:
Stop cleanly if a program being debugged spawns a domain

Otherwise the debugger gets all confused.

It is still possible to debug the program up to the point it spawns a
new domain.  This is the same solution we used in #10594 for programs
that create threads.

xavierleroy reviewed (dismissed):
Looks good to me.  Thanks a lot for this work.

I was able to run the demo program for effects from the OCaml 5 manual.  It's fun to step through all these continuations...

I added a commit that stops the debuggee cleanly if it spawns a domain.  This way we can still debug up to the point where a domain is created.  This is like #10594.

fabbing commented:
A bit of code may have been lost during the PR and this should probably be added because the trap barrier doesn't seem to work anymore:
```diff
diff --git a/runtime/debugger.c b/runtime/debugger.c
index 187f681135..a220915784 100644
--- a/runtime/debugger.c
+++ b/runtime/debugger.c
@@ -560,6 +560,8 @@ void caml_debugger(enum event_kind event, value param)
       caml_flush(dbg_out);
       break;
     case REQ_SET_TRAP_BARRIER:
+      i = caml_getword(dbg_in);
+      CAMLassert(i == frame_block_number(frame_block));
       i = caml_getword(dbg_in);
       Caml_state->trap_barrier_off = -i;
       break;
```

I noticed the problem, thanks to @OlivierNicole, while trying to review this PR (I'm very new to bytecode and ocamldebug).

However, I don't think it's enough and `trap_barrier_off` should probably be a tuple `{ block, offset }`.
Even a tuple may still cause issue if we switch to a different fiber at the same depth (`block`), in this case raising an exception could trigger a trap barrier into the wrong fiber.

damiendoligez commented:
@fabbing Thanks for finding this problem. This breaks the `finish` debugger command. I'm working on a solution.

damiendoligez committed:
fix the trap barrier and make it work with algebraic effects (aka fibers)

fabbing commented:
This is better than the "depth" in terms of number of fibers but will run into troubles if the fiber is reallocated since the `struct stack_info*` will change (see `caml_try_realloc_stack` from `runtime/fiber.c`).
As far as I know, there is currently no way to identify a fiber.
I was able to have a quick discussion on this topic with @kayceesrk and it seemed reasonable to add an unique identifier for each fiber. I'm working on a PR for this, I hope to have it ready soon.

damiendoligez committed:
use Caml_inline instead of inline

damiendoligez committed:
debugger: use fiber id instead of address

damiendoligez commented:
Adapted to use fiber IDs instead of addresses.
Another possibility would be to have `caml_try_realloc_stack` update `Caml_state->trap_barrier_block` as needed.

damiendoligez committed:
debugger: update .depend

fabbing commented:
I've resumed reviewing the PR and it looked good to me.
But I noticed that the `tool-debugger` tests from the testsuite where still skipped and tried to re-enable them again. All but one pass. The `dynlink` test fails with `Fatal error: Fatal error during unlock: Operation not permitted`.

It's not straightforward to debug such test as many fork are involved in the process.
I was able to pinpoint the issue at `channel_mutex_unlock_default` while attempting to create a `checkpoint` (`debugger.c:485`).
I don't understand why this only happens in this test and keep investigating!

gasche commented:
With the OCaml 5 runtime you cannot safely call `fork` from C and then execute any OCaml code on the child that would release the runtime lock, you have to do a special dance first (and meditate on the sins of doing anything else than running an exec* function). The error I encountered in this situation is similar to yours, maybe look at #11111 to see if it is related; in your stead I would look at uses of `fork` in `runtime/debugger.c`.

gasche commented on runtime/debugger.c:
```diff
@@ -440,7 +481,13 @@ void caml_debugger(enum event_kind event, value param)
       i = fork();
+      Lock (dbg_in);
+      Lock (dbg_out);
+      caml_acquire_domain_lock ();
```
Comparing this code with `caml_atfork_default` in domain.c, maybe you need to call `caml_reset_domain_lock` before re-acquiring the lock?

fabbing commented:
Thanks a lot for the hint @gasche. I had a look in that direction but it turns out to be something different:

This test raises an exception (`Not_found` from `Sys.getenv`) which calls `Unlock_exn` / `channel_mutex_unlock_exn_default` and since `dbg_out` channel was locked last, causes it to be unlocked.
Later, when trying to create a `checkpoint`, an attempt is made to unlock `dbg_out` before forking which fails since it's no longer locked.

A not very graceful solution is to set `last_channel_locked` to `NULL` to avoid this:
```diff
--- a/runtime/caml/io.h
+++ b/runtime/caml/io.h
@@ -115,6 +115,8 @@ CAMLextern struct channel * caml_all_opened_channels;
 #define Flush_if_unbuffered(channel) \
   if (channel->flags & CHANNEL_FLAG_UNBUFFERED) caml_flush(channel)
 
+CAMLextern void caml_channel_forget_last(void);
+
 /* Conversion between file_offset and int64_t */
 
 #define Val_file_offset(fofs) caml_copy_int64(fofs)
diff --git a/runtime/debugger.c b/runtime/debugger.c
index 861c37e3bb..448098e2fb 100644
--- a/runtime/debugger.c
+++ b/runtime/debugger.c
@@ -143,6 +143,7 @@ static void open_connection(void)
      private to this code and never used concurrently. */
   Lock (dbg_in);
   Lock (dbg_out);
+  caml_channel_forget_last();
 
   if (!caml_debugger_in_use) caml_putword(dbg_out, -1); /* first connection */
 #ifdef _WIN32
@@ -479,6 +480,7 @@ void caml_debugger(enum event_kind event, value param)
       i = fork();
       Lock (dbg_in);
       Lock (dbg_out);
+      caml_channel_forget_last();
       caml_acquire_domain_lock ();
       if (i == 0) {
         close_connection();     /* Close parent connection. */
diff --git a/runtime/io.c b/runtime/io.c
index 78bfea6d8a..fae4d6239e 100644
--- a/runtime/io.c
+++ b/runtime/io.c
@@ -106,6 +106,11 @@ CAMLexport void (*caml_channel_mutex_unlock_exn) (void)
 /* List of opened channels */
 CAMLexport struct channel * caml_all_opened_channels = NULL;
 
+CAMLexport void caml_channel_forget_last(void)
+{
+  last_channel_locked = NULL;
+}
+
 /* Basic functions over type struct channel *.
    These functions can be called directly from C.
    No locking is performed. */
```
Thanks again to @OlivierNicole for his help in understanding channel behaviour.

Unfortunately, this `dynlink` test still fails because unable to place a breakpoint in `Plugin`:
```
Can't find any event there.
```
Investigation continues!

damiendoligez committed:
re-enable debugger tests

damiendoligez committed:
Reset `last_channel_locked` after locking debugger's channels.
Patch written by @fabbing, slightly adapted by Damien Doligez.

damiendoligez commented:
This is strange. It says all tests have passed, but on my machine the dynlink test still fails.

damiendoligez committed:
re-enable dynlink debugging

damiendoligez commented:
>Can't find any event there.

Fixed by 3b34535a65a675bfa0bbf6ce2b2f883973cff5b7, let's see if the testsuite passes now.

fabbing commented:
Do we also want a "symmetrical" behavior with `resume` that would trigger a `trap` like `perform`?
```diff
--- a/runtime/interp.c
+++ b/runtime/interp.c
@@ -1284,6 +1284,7 @@ value caml_interprete(code_t prog, asize_t prog_size)
       goto do_resume;
 
 do_resume: {
+      check_trap_barrier_for_effect(domain_state);
       struct stack_info* stk = Ptr_val(accu);
       if (stk == NULL) {
          accu = Field(caml_global_data, CONTINUATION_ALREADY_TAKEN_EXN);
```
Either way it looks good to me!

damiendoligez commented:
>Do we also want a "symmetrical" behavior with `resume` that would trigger a trap like `perform`?

I don't think so. `resume` is like a function call, it doesn't break the control flow of the current function. The trap barrier is here to make sure we intercept all the ways to get out of the current function, not just by returning but also with an exception or an effect. If we wanted to be exhaustive, we should add a trap barrier to the `exit` system call (and also `kill` with an unhandled signal on the current process) but that's probably not worth the trouble.

damiendoligez commented:
@gasche, @xavierleroy can we get an official review ?

xavierleroy commented:
> can we get an official review ?

Yes, but please leave me 1-2 weeks to work on it.  I'm thinking about a radical solution to the locking problems that have plagued this PR.  Stay tuned.

xavierleroy committed:
Remove unnecessary locking on I/O channels used by the debugger

They are used in a single-threaded manner.
Locking is no longer necessary since #11356.

Also remove the recently-added caml_channel_reset_last_locked() function
which is no longer needed.

xavierleroy commented:
Here is a preview of the "radical solution" I propose: https://github.com/damiendoligez/ocaml/pull/7

Ideally, I should make a separate PR with the basic changes, have it merged, and then you would rebase this PR on top of this work.  But first I'd like you (@damiendoligez , @fabbing , ...) to have a quick look at the approach and comment.

xavierleroy committed:
Changes entry for #11065

xavierleroy commented:
I took the liberty to rebase on trunk and hand-merge https://github.com/damiendoligez/ocaml/pull/7.  I'm running a CI precheck test.

damiendoligez commented:
precheck is at https://ci.inria.fr/ocaml/job/precheck/726/

xavierleroy reviewed (approved):
Test results are as positive as they can be, and the whole PR is in good shape as far as I can see, so let's merge!  Thanks @damiendoligez for all the work.

xavierleroy commented:
Cherry-picked in 5.0: b4a61f5cba
